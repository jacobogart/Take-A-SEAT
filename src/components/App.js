import React, { Component } from 'react';
import AcronymContainer from './AcronymContainer';
import KeywordContainer from './KeywordContainer';
import Chalkboard from './Chalkboard';
import newLineChecks from '../newLineChecks';

class App extends Component {
  constructor() {
    super();
    this.state = {
      showKeywords: false,
      keywords: null,
      currentPhrase: "",
      stages: null,
      nextWords: null,
      chalkboardPhrases: [],
      chalkboardLines: 0
    };
  };

  componentDidMount() {
    const url = 'https://fe-apps.herokuapp.com/api/v1/memoize/1901/jacobogart/'
    const endPoints = ['stages', 'keywords']
    endPoints.forEach(endPoint => {
      fetch(`${url}${endPoint}`)
        .then(response => response.json())
        .then(data => this.setState({ [endPoint]: data[endPoint] }))
        .catch(err => console.log(err));
    })
  }


  findNextWords = keywordData => {
    let shallowNextWords = keywordData.nextWords || keywordData.starterWords;
    if (keywordData.word === "New Line") {
      shallowNextWords = this.newLineTestRunner(shallowNextWords, newLineChecks);
    }
    let newNextWords = shallowNextWords.map(nextWord =>
      this.state.keywords.find(keyword => keyword.word === nextWord)
    );
    
    this.setState({
      nextWords: newNextWords,
      showKeywords: true
    });
  };

  chalkboardChecker = keywordValue => {
    if(keywordValue === "New Line") {
      this.newChalkboardLine();
    } else {
      this.addToChalkboard(keywordValue);
    }
  };

  newChalkboardLine = () => {
    let newPhrase = this.state.chalkboardPhrases;
    newPhrase.push(this.state.currentPhrase);
    this.setState({
      chalkboardPhrases: newPhrase,
      currentPhrase: "",
      chalkboardLines: this.state.chalkboardLines + 1
    });
  };

  newLineTestRunner = (nextWords, newLineChecks) => {
    let workingNextWords = nextWords;
    newLineChecks.forEach(check => {
      let {find, ifFound, remove } = check;
      workingNextWords = this.newLineTest(workingNextWords, find, ifFound, remove); 
    });              
    return workingNextWords;
  };

  newLineTest = (workingNextWords, targetPhrase, boolean, removePhrase) => {
    let localNextwords = workingNextWords;
    if (this.newLineSome(targetPhrase) === boolean) {
      var index = localNextwords.indexOf(removePhrase);
      index >= 0 && localNextwords.splice(index, 1);
    } 
    return localNextwords;
  }

  newLineSome = (targetPhrase) => {
    return this.state.chalkboardPhrases.some(
      chalkPhrase =>
        chalkPhrase.includes(targetPhrase)) ||
      this.state.currentPhrase.includes(targetPhrase);
  };

  addToChalkboard = keywordValue => {
    let newPhrase = this.state.currentPhrase
      ? this.state.currentPhrase.concat("", keywordValue)
      : keywordValue;
    this.setState({
      currentPhrase: newPhrase
    });
  };

  render() {
    return (
      <main className="App">
        {!this.state.showKeywords && 
          <h2 className="welcome-message">Take a</h2>
        }
        <AcronymContainer
          stages={this.state.stages}
          findNextWords={this.findNextWords}
          showKeywords={this.state.showKeywords}
        />
        {this.state.showKeywords && (
          <KeywordContainer
            keywords={this.state.keywords}
            chalkboardChecker={this.chalkboardChecker}
            nextWords={this.state.nextWords}
            findNextWords={this.findNextWords}
          />
        )}
        {this.state.showKeywords && (
          <Chalkboard
            currentPhrase={this.state.currentPhrase}
            chalkboardPhrases={this.state.chalkboardPhrases}
          />
        )}
      </main>
    );
  }
}

export default App;
