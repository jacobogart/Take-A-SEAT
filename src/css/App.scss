.App {
  text-align: center;
  display: grid;
  grid-template-columns: 400px 1fr;
  grid-template-rows: 200px 1fr;
  grid-template-areas: 
  "acro acro"
  "keywords chalkboard";
  height: 740px;
  width: 100%;
}

