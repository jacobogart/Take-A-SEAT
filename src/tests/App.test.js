import React from 'react';
import App from '../components/App';
import { shallow } from "enzyme";

const defaultState = {
  showKeywords: false,
  keywords: null,
  currentPhrase: "",
  stages: null,
  nextWords: null,
  chalkboardPhrases: [],
  chalkboardLines: 0
};

const mockKeywords = [
  {
    "word": "Prop Name",
    "nextWords": [
      "mockData Name",
      "mockPropFunction"
    ],
    "value": "yourPropName={",
    "isEditable": true,
    "phase": ["S"],
    "details": "This is how you will access the prop's value in the child component",
    "id": 114
    },
    {
    "word": "mockData Name",
    "nextWords": [
      "=",
      "}"
    ],
    "value": "mock-",
    "isEditable": true,
    "phase": ["S"],
    "details": "Variable name for your mock data",
    "id": 101
    },
    {
    "word": "mockPropFunction",
    "nextWords": [
      "=",
      "}"
    ],
    "value": "mock-",
    "isEditable": true,
    "phase": ["S"],
    "details": "Name of mock function you will pass as props and/or spy on",
    "id": 102
  },
  {
    "word": "New Line",
    "nextWords": [
      ")",
      "let",
      "wrapper",
      "expect",
      "beforeEach",
      "Prop Name",
      "/>",
      "}",
      "describe",
      "Component Name",
      "it"
    ],
    "value": "New Line",
    "isEditable": false,
    "phase": ["S", "E", "A", "T"],
    "details": "Start a new line of testing",
    "id": 127
  },

]
;

describe('App', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <App />
    );
  });
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a default state', () => {
    expect(wrapper.state()).toEqual(defaultState);
  });

  it('should render KeywordContainer and Chalkboard if showKeywords is true', () => {
    wrapper.instance().setState({ showKeywords: true });
    expect(wrapper).toMatchSnapshot();
  });

  describe('findNextWords', () => {
    it('should set nextWords to an array of keyword objects', () => {
      wrapper.instance().setState({ keywords: mockKeywords })
      wrapper.instance().findNextWords(mockKeywords[0]);
      expect(wrapper.state('nextWords')).toEqual([mockKeywords[1], mockKeywords[2]]);
    });

    it('should update showKeywords to true', () => {
      wrapper.instance().setState({ keywords: mockKeywords })
      wrapper.instance().findNextWords(mockKeywords[0]);
      expect(wrapper.state('showKeywords')).toEqual(true);
    });
    it("should invoke newLineTests if the keyword is New Line", () => {
      wrapper.instance().setState({ keywords: mockKeywords });
      jest.spyOn(wrapper.instance(), "newLineTests");
      wrapper.instance().findNextWords(mockKeywords[3]);
      expect(wrapper.instance().newLineTests).toHaveBeenCalled();
    });
  });

  describe('chalkboardChecker', () => {
    it("should invoke newChalkboardLine if keyboardValue is New Line", () => {
      let instance = wrapper.instance(); 
      jest.spyOn(instance, "newChalkboardLine");
      instance.chalkboardChecker('New Line');
      expect(instance.newChalkboardLine).toHaveBeenCalled();
    });

    it("should invoke addToChalkboard if keyboardValue is not New Line", () => {
      let instance = wrapper.instance();
      jest.spyOn(instance, "addToChalkboard");
      instance.chalkboardChecker("=");
      expect(instance.addToChalkboard).toHaveBeenCalled();
      expect(instance.addToChalkboard).toHaveBeenCalledWith("=");
    });
  });

  describe('newChalkboardLine', () => {
    it('should update chalkboardPhrases when newChalkboardLine is called', () => {
      let instance = wrapper.instance();
      expect(wrapper.state()).toEqual(defaultState);
      instance.setState({
        currentPhrase: "mock current phrase"
      });
      instance.newChalkboardLine();
      expect(wrapper.state("chalkboardPhrases")).toEqual([
        "mock current phrase"
      ]);
    });

    it("should update currentPhrase when newChalkboardLine is called", () => {
      let instance = wrapper.instance();
      expect(wrapper.state()).toEqual(defaultState);
      instance.setState({
        currentPhrase: "mock current phrase"
      });
      expect(wrapper.state("currentPhrase")).toEqual("mock current phrase");
      instance.newChalkboardLine();
      expect(wrapper.state("currentPhrase")).toEqual('');
    });
  });

  describe('addToChalkboard', () => {
    it('should add keywordValue to the current string, if there is already a currentPhrase', () => {
      let instance = wrapper.instance();
      expect(wrapper.state()).toEqual(defaultState);
      instance.setState({
        currentPhrase: "mock current phrase"
      });
      instance.addToChalkboard(" test");
      expect(wrapper.state('currentPhrase')).toEqual("mock current phrase test");
    });

    it("should set currentPhrase to keywordValue if there isn't a currentPhrase", () => {
      let instance = wrapper.instance();
      expect(wrapper.state()).toEqual(defaultState);
      instance.addToChalkboard("test");
      expect(wrapper.state("currentPhrase")).toEqual("test");
    });
  });
});
