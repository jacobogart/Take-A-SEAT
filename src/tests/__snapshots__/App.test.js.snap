// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<main
  className="App"
>
  <h2
    className="welcome-message"
  >
    Take a
  </h2>
  <AcronymContainer
    findNextWords={[Function]}
    showKeywords={false}
    stages={null}
  />
</main>
`;

exports[`App should render KeywordContainer and Chalkboard if showKeywords is true 1`] = `
<main
  className="App"
>
  <AcronymContainer
    findNextWords={[Function]}
    showKeywords={true}
    stages={null}
  />
  <KeywordContainer
    chalkboardChecker={[Function]}
    findNextWords={[Function]}
    keywords={null}
    nextWords={null}
  />
  <Chalkboard
    chalkboardPhrases={Array []}
    currentPhrase=""
  />
</main>
`;
